File: adapter/harness/investigation.py
"""
Bounded read-only investigation sub-agent — S4 of
plans/harness_trajectory_supervisor_plan.html (GATHER_EVIDENCE).

When the Trajectory Supervisor decides the run is stuck for lack of a fact, it
emits a ``GATHER_EVIDENCE`` directive carrying an ``InvestigationRequest``. This
module runs that request as a *plain, bounded, read-only tool loop* — deliberately
NOT a recursive ``run_one_iteration`` (Q2 = bounded tool loop for v1). The findings
are merged back into the parent ``world_model`` as provenanced observations and the
generation id is bumped so staleness + contradiction detection re-run over them.

Invariants enforced here (CI-gated from S4):
  INV-23  Investigation sub-agents have no write / shell / email tools — the
          ``suggested_tools`` list is filtered to INVESTIGATION_READ_ONLY_TOOLS
          before any dispatch; a write/shell/email/unknown fn_ref is rejected.
  INV-24  Investigation depth is capped at 1 — ``run_investigation(depth>=1)``
          raises, and a GATHER_EVIDENCE directive surfaced from inside an
          investigation context is coerced away by the caller.
  INV-25  Every investigation runs under its own bounded call budget, independent
          of the parent RecoveryBudget; exhaustion returns partial findings and
          degrades the directive — it never hangs (per-call bounded timeout).
"""

from __future__ import annotations

import uuid
from collections.abc import Callable
from concurrent.futures import ThreadPoolExecutor
from concurrent.futures import TimeoutError as _FutureTimeout
from dataclasses import dataclass, field
from typing import Any

from .staleness import increment_generation_id
from .world_model import Observation, WorldModel

# Read-only fn_ref subset an investigation may touch. A tool not in this set —
# including every write / shell / email tool — is rejected before dispatch (INV-23).
INVESTIGATION_READ_ONLY_TOOLS: frozenset[str] = frozenset(
    {
        "retrieve",
        "rag_retrieve",
        "kb_retrieve",
        "search",
        "web_search",
        "fetch_url",
        "http_get",
        "read_file",
        "list_directory",
        "get_session_state",
        "list_reminders",
        "lookup",
    }
)

# Explicitly named for the INV-23 test / logging clarity; membership in the
# allowlist above is what actually gates dispatch.
INVESTIGATION_FORBIDDEN_TOOLS: frozenset[str] = frozenset(
    {"write_file", "run_shell_command", "shell", "exec", "send_email", "deploy", "apply_patch"}
)

_MAX_DEPTH = 1
_MAX_CALL_BUDGET = 20
_DEFAULT_PER_CALL_TIMEOUT = 15.0
_MAX_FINDING_LEN = 800


class InvestigationDepthExceeded(RuntimeError):
    """Raised when run_investigation is called at depth >= 1 (INV-24)."""


def _clip(text: Any, limit: int = _MAX_FINDING_LEN) -> str:
    s = str(text or "").strip()
    return s if len(s) <= limit else s[: limit - 1] + "…"


def _call_tool_bounded(tool_runner: Callable[[str, str], Any], tool: str, question: str, timeout: float) -> Any:
    """One read-only tool call under a hard per-call timeout (INV-25). A call that
    raises, or hangs past ``timeout``, yields ``None`` — never propagates. Each call
    gets its own single-thread executor so a hung call can't wedge the next tool; the
    straggler thread is abandoned (shutdown(wait=False)) and dies with the process."""
    pool = ThreadPoolExecutor(max_workers=1)
    try:
        return pool.submit(tool_runner, tool, question).result(timeout=timeout)
    except _FutureTimeout:
        return None
    except Exception:
        return None
    finally:
        pool.shutdown(wait=False, cancel_futures=True)


@dataclass
class InvestigationFinding:
    content: str
    tool: str
    reliability: str = "MEDIUM"  # "MEDIUM" | "HIGH"


@dataclass
class InvestigationOutcome:
    findings: list[InvestigationFinding] = field(default_factory=list)
    calls_made: int = 0
    exhausted: bool = False
    rejected_tools: list[str] = field(default_factory=list)

    @property
    def is_empty(self) -> bool:
        return not self.findings


def validate_investigation_tools(tools: list[str]) -> tuple[list[str], list[str]]:
    """Split a suggested-tool list into (allowed, rejected). INV-23."""
    allowed: list[str] = []
    rejected: list[str] = []
    for raw in tools or []:
        name = str(raw or "").strip()
        if name and name in INVESTIGATION_READ_ONLY_TOOLS:
            allowed.append(name)
        elif name:
            rejected.append(name)
    return allowed, rejected


def run_investigation(
    request: Any,
    *,
    tool_runner: Callable[[str, str], Any],
    tool_reliability: dict[str, str] | None = None,
    depth: int = 0,
    per_call_timeout: float = _DEFAULT_PER_CALL_TIMEOUT,
) -> InvestigationOutcome:
    """Run one bounded read-only investigation.

    ``request`` is an ``InvestigationRequest`` (question / suggested_tools / budget).
    ``tool_runner(tool_name, question) -> str | None`` is host-supplied and does the
    actual (read-only) work. Never raises for tool faults — a tool that raises,
    returns nothing, or hangs past ``per_call_timeout`` is skipped and counted.
    """
    if depth >= _MAX_DEPTH:
        raise InvestigationDepthExceeded(
            f"investigation depth {depth} >= {_MAX_DEPTH}: an investigation cannot spawn an investigation (INV-24)"
        )

    question = _clip(getattr(request, "question", ""), 600)
    suggested = list(getattr(request, "suggested_tools", []) or [])
    raw_budget = int(getattr(request, "budget", 5) or 0)
    budget = max(0, min(raw_budget, _MAX_CALL_BUDGET))

    allowed, rejected = validate_investigation_tools(suggested)
    outcome = InvestigationOutcome(rejected_tools=rejected)
    if not question or not allowed or budget == 0:
        outcome.exhausted = bool(allowed) and budget == 0
        return outcome

    for tool in allowed:
        if outcome.calls_made >= budget:
            outcome.exhausted = True
            break
        outcome.calls_made += 1
        result = _call_tool_bounded(tool_runner, tool, question, per_call_timeout)
        text = _clip(result)
        if not text:
            continue
        reliability = "HIGH" if (tool_reliability or {}).get(tool) == "HIGH" else "MEDIUM"
        outcome.findings.append(InvestigationFinding(content=text, tool=tool, reliability=reliability))

    if not outcome.exhausted and outcome.calls_made < len(allowed):
        outcome.exhausted = True
    return outcome


def merge_investigation_findings(
    world_model: WorldModel,
    outcome: InvestigationOutcome,
    *,
    question: str,
) -> list[Observation]:
    """Merge investigation findings into the parent world_model as provenanced
    observations (source="supervisor_investigation"), then bump the generation id
    so staleness + contradiction detection re-run over the new evidence.

    The Python ``Observation`` has no ``derived_from`` field (INV-01 is a
    belief-level invariant — see the S3 note); provenance here is carried by
    ``source`` plus an explicit prefix in ``content``.
    """
    merged: list[Observation] = []
    q = _clip(question, 300).replace("]", " ")
    for finding in outcome.findings:
        obs = Observation(
            id=f"supervisor_investigation:{uuid.uuid4().hex[:12]}",
            content=(
                f"[supervisor_investigation q={q}] "
                f"[tool={finding.tool} reliability={finding.reliability} "
                f"derived_from=supervisor_investigation] {finding.content}"
            ),
            source="supervisor_investigation",
        )
        world_model.add_observation(obs)
        merged.append(obs)
    increment_generation_id(world_model)
    return merged


def count_investigations(world_model: WorldModel) -> int:
    """Distinct investigations already merged this run — used for the per-run cap K.

    Counts unique ``q=...`` prefixes among supervisor_investigation observations so
    multiple findings from one investigation count once.
    """
    seen: set[str] = set()
    for obs in getattr(world_model, "observations", []) or []:
        if obs.source == "supervisor_investigation" and obs.content.startswith("[supervisor_investigation "):
            head = obs.content.split("]", 1)[0]
            seen.add(head)
    return len(seen)

File: packages/harness/src/investigation.ts
// Bounded read-only investigation sub-agent — S5 of
// plans/harness_trajectory_supervisor_plan.html (GATHER_EVIDENCE). TS twin of
// adapter/harness/investigation.py.
//
// When the Trajectory Supervisor emits a GATHER_EVIDENCE directive at a stall edge,
// the harness runs a bounded read-only investigation (via a host-provided callback —
// see HarnessRunOptions.runInvestigation) and merges the findings back into the
// parent WorldModel as provenanced observations, bumping the generation id so
// staleness + contradiction detection re-run over them.
//
// Invariants (mirrors the Python module, CI-gated from S4/S5):
//   INV-23  Investigation sub-agents have no write / shell / email tools —
//           suggested_tools is filtered to INVESTIGATION_READ_ONLY_TOOLS before any
//           dispatch (validateInvestigationTools).
//   INV-24  Investigation depth is capped at 1 — runInvestigation(depth>=1) throws.
//   INV-25  Every investigation runs under its own bounded call budget; a tool that
//           hangs past perCallTimeoutMs is abandoned and the run still returns.

import type { Observation, WorldModel } from './state/world-model.js'
import { SupervisorDirective, type InvestigationRequestData } from './supervisor.js'

/** Read-only fn_ref subset an investigation may touch. A tool not in this set —
 *  including every write / shell / email tool — is rejected before dispatch (INV-23).
 *  Kept byte-identical to investigation.py's INVESTIGATION_READ_ONLY_TOOLS. */
export const INVESTIGATION_READ_ONLY_TOOLS: ReadonlySet<string> = new Set([
  'retrieve',
  'rag_retrieve',
  'kb_retrieve',
  'search',
  'web_search',
  'fetch_url',
  'http_get',
  'read_file',
  'list_directory',
  'get_session_state',
  'list_reminders',
  'lookup',
])

/** Named for clarity / logging; membership in the allowlist above is what gates dispatch. */
export const INVESTIGATION_FORBIDDEN_TOOLS: ReadonlySet<string> = new Set([
  'write_file',
  'run_shell_command',
  'shell',
  'exec',
  'send_email',
  'deploy',
  'apply_patch',
])

const MAX_DEPTH = 1
const MAX_CALL_BUDGET = 20
const DEFAULT_PER_CALL_TIMEOUT_MS = 15_000
const MAX_FINDING_LEN = 800

export class InvestigationDepthExceededError extends Error {
  constructor(depth: number) {
    super(
      `investigation depth ${depth} >= ${MAX_DEPTH}: an investigation cannot spawn an investigation (INV-24)`,
    )
    this.name = 'InvestigationDepthExceededError'
  }
}

function clip(text: unknown, limit = MAX_FINDING_LEN): string {
  const s = String(text ?? '').trim()
  return s.length <= limit ? s : s.slice(0, limit - 1) + '…'
}

export interface InvestigationFinding {
  content: string
  tool: string
  reliability: 'MEDIUM' | 'HIGH'
}

export interface InvestigationOutcome {
  findings: InvestigationFinding[]
  callsMade: number
  exhausted: boolean
  rejectedTools: string[]
}

export interface InvestigationRequestLike {
  question?: string
  suggested_tools?: string[]
  budget?: number
}

/** A host-supplied read-only tool call: (toolName, question) → result string | null.
 *  May be sync or async. Faults / hangs are isolated by runInvestigation. */
export type InvestigationToolRunner = (tool: string, question: string) => string | null | Promise<string | null>

/** Split a suggested-tool list into { allowed, rejected }. INV-23. Byte-identical to
 *  investigation.py's validate_investigation_tools(). */
export function validateInvestigationTools(tools: readonly string[] | undefined): {
  allowed: string[]
  rejected: string[]
} {
  const allowed: string[] = []
  const rejected: string[] = []
  for (const raw of tools ?? []) {
    const name = String(raw ?? '').trim()
    if (!name) continue
    if (INVESTIGATION_READ_ONLY_TOOLS.has(name)) allowed.push(name)
    else rejected.push(name)
  }
  return { allowed, rejected }
}

async function callToolBounded(
  toolRunner: InvestigationToolRunner,
  tool: string,
  question: string,
  timeoutMs: number,
): Promise<string | null> {
  // A hung tool call must not wedge the loop (INV-25). Promise.race against a timer;
  // a rejection or timeout yields null rather than propagating. The straggler promise
  // is abandoned.
  let timer: ReturnType<typeof setTimeout> | undefined
  const timeout = new Promise<null>(resolve => {
    timer = setTimeout(() => resolve(null), timeoutMs)
  })
  try {
    const result = await Promise.race([Promise.resolve().then(() => toolRunner(tool, question)), timeout])
    return result ?? null
  } catch {
    return null
  } finally {
    if (timer) clearTimeout(timer)
  }
}

/** Run one bounded read-only investigation. `depth` >= 1 throws (INV-24). Never
 *  rejects for tool faults — a tool that throws, returns nothing, or hangs past
 *  `perCallTimeoutMs` is skipped and counted (INV-25). */
export async function runInvestigation(
  request: InvestigationRequestLike,
  opts: {
    toolRunner: InvestigationToolRunner
    toolReliability?: Record<string, string>
    depth?: number
    perCallTimeoutMs?: number
  },
): Promise<InvestigationOutcome> {
  const depth = opts.depth ?? 0
  if (depth >= MAX_DEPTH) throw new InvestigationDepthExceededError(depth)

  const question = clip(request.question ?? '', 600)
  const rawBudget = Number(request.budget ?? 5)
  const budget = Math.max(0, Math.min(Number.isFinite(rawBudget) ? Math.trunc(rawBudget) : 5, MAX_CALL_BUDGET))
  const { allowed, rejected } = validateInvestigationTools(request.suggested_tools)

  const outcome: InvestigationOutcome = { findings: [], callsMade: 0, exhausted: false, rejectedTools: rejected }
  if (!question || allowed.length === 0 || budget === 0) {
    outcome.exhausted = allowed.length > 0 && budget === 0
    return outcome
  }

  const timeoutMs = opts.perCallTimeoutMs ?? DEFAULT_PER_CALL_TIMEOUT_MS
  for (const tool of allowed) {
    if (outcome.callsMade >= budget) {
      outcome.exhausted = true
      break
    }
    outcome.callsMade++
    const raw = await callToolBounded(opts.toolRunner, tool, question, timeoutMs)
    const text = clip(raw)
    if (!text) continue
    const reliability = opts.toolReliability?.[tool] === 'HIGH' ? 'HIGH' : 'MEDIUM'
    outcome.findings.push({ content: text, tool, reliability })
  }

  if (!outcome.exhausted && outcome.callsMade < allowed.length) outcome.exhausted = true
  return outcome
}

let _obsCounter = 0

/** Merge investigation findings into the parent WorldModel as provenanced
 *  observations (source="supervisor_investigation"), then bump the generation id so
 *  staleness + contradiction detection re-run. Mirrors merge_investigation_findings()
 *  in the Python twin (Observation has no derived_from field — INV-01 is belief-level;
 *  provenance is `source` plus a content prefix). */
export function mergeInvestigationFindings(
  worldModel: WorldModel,
  findings: readonly InvestigationFinding[],
  meta: { question: string },
): Observation[] {
  const q = clip(meta.question, 300).replace(/]/g, ' ')
  const merged: Observation[] = []
  for (const finding of findings) {
    const obs: Observation = {
      id: `supervisor_investigation:${(++_obsCounter).toString(36)}${Date.now().toString(36)}`,
      content:
        `[supervisor_investigation q=${q}] ` +
        `[tool=${finding.tool} reliability=${finding.reliability} ` +
        `derived_from=supervisor_investigation] ${finding.content}`,
      source: 'supervisor_investigation',
      recorded_at: new Date().toISOString(),
    }
    worldModel.observations.push(obs)
    merged.push(obs)
  }
  worldModel.incrementGenerationId()
  return merged
}

/** Distinct investigations already merged this run — used for the per-run cap K.
 *  Counts unique `[supervisor_investigation q=…]` prefixes so multiple findings from
 *  one investigation count once. Mirrors count_investigations() in the Python twin. */
export function countInvestigations(worldModel: WorldModel): number {
  const seen = new Set<string>()
  for (const obs of worldModel.observations ?? []) {
    if (obs.source === 'supervisor_investigation' && obs.content.startsWith('[supervisor_investigation ')) {
      seen.add(obs.content.split(']', 1)[0])
    }
  }
  return seen.size
}

/** Per-run cap K on supervisor investigations (Q4). Beyond it, GATHER_EVIDENCE
 *  degrades — to ASK_USER once wired, to CONTINUE today. Same value as loop.py's
 *  _SUPERVISOR_INVESTIGATION_CAP. */
export const INVESTIGATION_CAP_K = 3

/** A host-provided bounded read-only investigation: request → findings. The host
 *  runs its own tool loop (own Budget, tool-policy gate). Absent → GATHER_EVIDENCE
 *  degrades to CONTINUE. */
export type RunInvestigationHost = (req: InvestigationRequestData) => Promise<InvestigationFinding[]>

/**
 * Resolve a GATHER_EVIDENCE directive at a stall edge (S5). Runs the host
 * investigation, merges its findings into `worldModel` with provenance + a
 * generation bump, and returns a coerced CONTINUE so the deterministic ladder
 * proceeds over the new evidence. Degrades to CONTINUE — carrying a labelled
 * rationale — when: no host callback, per-run cap K already hit, or the host
 * throws / rejects. Never throws.
 *
 * loop.py (S4, Python) does the equivalent split across run_one_iteration()'s
 * early-return + the driver's run_investigation()/merge_investigation_findings();
 * the two must stay behaviourally identical for the degradation + cap rules.
 */
export async function resolveGatherEvidence(
  worldModel: WorldModel,
  directive: SupervisorDirective,
  host: RunInvestigationHost | undefined,
): Promise<SupervisorDirective> {
  if (!host || !directive.investigation) {
    return SupervisorDirective.cont(`[not wired: GATHER_EVIDENCE] ${directive.rationale}`)
  }
  if (countInvestigations(worldModel) >= INVESTIGATION_CAP_K) {
    return SupervisorDirective.cont(`[investigation cap ${INVESTIGATION_CAP_K} reached] ${directive.rationale}`)
  }
  const req = directive.investigation.toJSON()
  let findings: InvestigationFinding[]
  try {
    findings = (await host(req)) ?? []
  } catch {
    return SupervisorDirective.cont(`[investigation failed] ${directive.rationale}`)
  }
  if (findings.length > 0) mergeInvestigationFindings(worldModel, findings, { question: req.question })
  return SupervisorDirective.cont(`[investigation done] ${directive.rationale}`)
}
